enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,243 @@
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `*ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  NameSuffix string `json:"nameSuffix"`
+  // DiskSizeGB is the size in GB to assign to the data disk.
+  DiskSizeGB int32 `json:"diskSizeGB"`
+  // ManagedDisk specifies the Managed Disk parameters for the data disk.
+  // +optional
+  ManagedDisk *ManagedDiskParameters `json:"managedDisk,omitempty"`
+  // Lun Specifies the logical unit number of the data disk. This value is used to identify data disks within the VM and therefore must be unique for each data disk attached to a VM.
+  // The value must be between 0 and 63.
+  // +optional
+  Lun *int32 `json:"lun,omitempty"`
+  // CachingType specifies the caching requirements.
+  // +optional
+  // +kubebuilder:validation:Enum=None;ReadOnly;ReadWrite
+  CachingType string `json:"cachingType,omitempty"`
```

Comment:
do you think it makes sense to add a wrapper around string here like this?
```suggestion
  CachingType CachingTypeOption `json:"cachingType,omitempty"`
```

and then you can have
```golang
const ReadOnly CachingTypeOption = "ReadOnly"
const ReadWrite CachingTypeOption = "ReadWrite"
```

Reply:
Yes I think that could make a lot of sense and it's potentially how I would to things if I were to do this from scratch. That said, I copied this from upstream and I could see the value of keeping it similar, but if we think it's better to we can do so 👍

Reply:
Either way, it is API compatible so I think I would probably go down the enum option even though it does differ from upstream, means we have constants to use when we are validating which is useful

Reply:
I see, I was wondering that! Will update, thanks.

Reply:
Ok, I've updated it and added also a third `CachingTypeOption` value to @alexander-demichev  suggestion for `"None"` to keep it consistent with the 3-state Enum.